src/gpio.rs

```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
-        impl<MODE> InputPin for PXx<Input<MODE>> {
-            type Error = Infallible;
+impl<GPIO, INDEX, OTYPE> OutputPin for Pin<GPIO, INDEX, Output<OTYPE>>
+where
+    GPIO: Gpio,
+    INDEX: Index,
+{
+    type Error = Infallible;
+
+    fn set_high(&mut self) -> Result<(), Self::Error> {
+        // NOTE(unsafe) atomic write to a stateless register
+        unsafe { (*self.gpio.ptr()).set_high(self.index.index()) };
+        Ok(())
+    }
+
+    fn set_low(&mut self) -> Result<(), Self::Error> {
+        // NOTE(unsafe) atomic write to a stateless register
+        unsafe { (*self.gpio.ptr()).set_low(self.index.index()) };
+        Ok(())
+    }
+}
 
-            fn is_high(&self) -> Result<bool, Self::Error> {
-                Ok(!self.is_low()?)
-            }
+#[cfg(feature = "unproven")]
+impl<GPIO, INDEX, MODE> InputPin for Pin<GPIO, INDEX, MODE>
+where
+    GPIO: Gpio,
+    INDEX: Index,
+    MODE: Readable,
+{
+    type Error = Infallible;
+
+    fn is_high(&self) -> Result<bool, Self::Error> {
+        Ok(!self.is_low()?)
+    }
+
+    fn is_low(&self) -> Result<bool, Self::Error> {
+        // NOTE(unsafe) atomic read with no side effects
+        Ok(unsafe { (*self.gpio.ptr()).is_low(self.index.index()) })
+    }
+}
 
-             fn is_low(&self) -> Result<bool, Self::Error> {
-                // NOTE(unsafe) atomic read with no side effects
-                Ok(unsafe {
-                    match &self.gpio {
-                        $(
-                            Gpio::$GPIOX => (*$GPIOX::ptr()).idr.read().bits() & (1 << self.i) == 0,
-                        )+
-                    }
-                })
-            }
-        }
+#[cfg(feature = "unproven")]
+impl<GPIO, INDEX, OTYPE> StatefulOutputPin for Pin<GPIO, INDEX, Output<OTYPE>>
+where
+    GPIO: Gpio,
+    INDEX: Index,
+{
+    fn is_set_high(&self) -> Result<bool, Self::Error> {
+        Ok(!self.is_set_low()?)
+    }
+
+    fn is_set_low(&self) -> Result<bool, Self::Error> {
+        // NOTE(unsafe) atomic read with no side effects
+        Ok(unsafe { (*self.gpio.ptr()).is_set_low(self.index.index()) })
+    }
+}
 
-        #[cfg(feature = "unproven")]
-        impl InputPin for PXx<Output<OpenDrain>> {
-            type Error = Infallible;
+#[cfg(feature = "unproven")]
+impl<GPIO, INDEX, OTYPE> toggleable::Default for Pin<GPIO, INDEX, Output<OTYPE>>
+where
+    GPIO: Gpio,
+    INDEX: Index,
+{
+}
 
-            fn is_high(&self) -> Result<bool, Self::Error> {
-                Ok(!self.is_low()?)
+impl<GPIO, INDEX, MODE> Pin<GPIO, INDEX, MODE>
+where
+    GPIO: Gpio,
+    INDEX: Index,
+    MODE: Active,
+{
+    /// Make corresponding EXTI line sensitive to this pin
+    pub fn make_interrupt_source(&mut self, syscfg: &mut SysCfg) {
+        let i = self.index.index() % 4;
+        let extigpionr = self.gpio.port_index() as u32;
+        match self.index.index() {
+            0..=3 => unsafe { modify_at!(syscfg.exticr1, 4, i, extigpionr) },
```

Comment:
Consider adding some let expression before the `modify_at` macro that labels what the number 4 is.